src/transform.js

```diff
@@ -0,0 +1,40 @@
+import { properties as childProperties } from "./children.js";
+import { matches } from "./util.js";
+
+
+/**
+ * Recursively execute a callback on this node and all its children.
+ * If the callback returns a non-undefined value, it will overwrite the node.
+ * 
+ * @param {object | object[]} node AST node or array of nodes
+ * @param {function(object, string, object?)} transformCallback
+ * @param {object} [o]
+ * @param {string | string[] | function} [o.only] Only walk nodes of this type
+ * @param {string | string[] | function} [o.except] Ignore walking nodes of these types
+ */
+export default function transform (node, transformCallback, o) {
+	_transform(node, transformCallback, o);
+}
+
+function _transform (node, callback, o = {}, property, parent) {
+	if (Array.isArray(node)) {
+		return node.forEach(n => _transform(n, callback, o, property, parent));
+	}
+
+	const ignore = o.except && matches(node, o.except);
+	const explore = !ignore && matches(node, o.only);
+	
+	if (explore) {
+		const transformedNode = callback(node, property, parent);
+		if (transformedNode !== undefined) {
+			// delete all keys in node:
+			Object.keys(node).forEach(key => delete node[key]);
+			// copy all keys from transformedNode to node:
+			Object.assign(node, transformedNode);
```

Comment:
This is what I have as a placeholder, deleting all of node's keys before reassigning, however, this is imperfect since it has the potential to delete something useful like `parent`. Is it the implementor's job to make sure they preserve `parent` or any other desired keys in the return from `callback`? For example, a callback a user would supply could be like:

```js
function callback(node, property, parent) {
    // some calculations ...
    return {parent, ...};
}
```

Reply:
@LeaVerou curious to see what you think of this. In my opinion, it's the client of this function's job to preserve any properties from `node` they want to keep, so deleting all node's keys and replacing them with transformedNode makes sense to me, but maybe I'm thinking about this wrong

Reply:
To the extent possible, if parent references exist, we need to reassign them correctly. Obviously we can't do all the work with complex transformations, but remember "simple things should be easy, complex things should be possible".

Reply:
Ok, if we move forward with this approach we can add a line that preserves the parent property

Reply:
@LeaVerou do we have to worry about this actually? Since [we define the parent property as non-enumerable](https://github.com/mavoweb/vastly/blob/main/src/parents.js#L38), `Object.keys` in theory should not include `parent`?

Reply:
Let me test it

Reply:
Seems like this is the case

Reply:
@LeaVerou I think we should be fine here, what do you think?

Reply:
I wasn't just talking about removing parents. If parent pointers are present, they should be readjusted. E.g. if the child gets replaced, it should have a parent pointer to its new parent. If the parent gets replaced, its children should also be adjusted. This makes me think, `parents.set()` should probably have a shallow version, when you don't want to descend to a whole subtree. What do you think?

Reply:
Definitely agree that parent pointers should be maintained consistently in here. I thought that due to aliasing, when a node gets modified, if it has children, their references will still be to the new node. I did a few tests to see if this was the case and it looked like it, but I can try to figure out a way to test for this